Fix highlight end pixel position calculation

nathansobo opened the pull request:
Previously, we were calculating the position preceding block decorations for the row following the end of the highlighted range, but that's actually wrong. We just want the position following block decorations of the end of the highlighted range, plus one line height. This prevents us from incorrectly rendering the end of highlight after block decorations that immediately follow the end of the highlighted range.

Before:

<img width="316" alt="screen shot 2017-08-18 at 8 56 51 pm" src="https://user-images.githubusercontent.com/1789/29482915-ded1c2cc-8457-11e7-8351-e954d1429aa2.png">

After:

<img width="362" alt="screen shot 2017-08-18 at 8 56 44 pm" src="https://user-images.githubusercontent.com/1789/29482921-f6ad596a-8457-11e7-9e4e-07ae2aba52ad.png">

/cc @ungb @as-cii

Commit:
Fix highlight end pixel position calculation

Previously, we were calculating the position preceding block decorations
for the row following the end of the highlighted range, but that's
actually wrong. We just want the position following block decorations of
the end of the highlighted range, plus one line height. This prevents us
from incorrectly rendering the end of highlight after block decorations
that immediately follow the end of the highlighted range.

ungb commented:
I was testing this on merge conflict decorators and found the following issues:

The highlighting from the `Our Changes` is off
![image](https://user-images.githubusercontent.com/5624870/29483184-c728fe80-8455-11e7-9e7d-21092476379e.png)


When you first open the merge conflict you can sometimes see the following:
![image](https://user-images.githubusercontent.com/5624870/29483195-e9aa4d2e-8455-11e7-9492-76cd60741cd4.png)


When you change the window size the cursor eventually will go to the right spot and the issue is gone and only come back when you close and open the merge window.

Ping me if you can't repro the issue.

Commit:
Handle highlight end rows with 'before' blocks in addition to 'after'

Commit:
Prevent block decoration margins from collapsing during measurement

We now render a 1px high sentinel element between off-screen block
decorations before measuring them to prevent margins from collapsing.

Commit:
Put back mistakenly deleted line

as-cii commented:
Nice work on this one, especially on figuring out the margin collapse issue. 💯 I'll go ahead and merge this once we get a green build.
